lib/handlers/token-handler.js

```diff
@@ -0,0 +1,249 @@
+    throw new InvalidArgumentError('Missing parameter: `model`');
+  }
+
+  if (!options.refreshTokenLifetime) {
+    throw new InvalidArgumentError('Missing parameter: `refreshTokenLifetime`');
+  }
+
+  if (!options.model.getClient) {
+    throw new InvalidArgumentError('Invalid argument: model does not implement `getClient()`');
+  }
+
+  this.accessTokenLifetime = options.accessTokenLifetime;
+  this.grantTypes = _.assign({}, grantTypes, options.extendedGrantTypes);
+  this.model = options.model;
+  this.refreshTokenLifetime = options.refreshTokenLifetime;
+}
+
+/**
+ * Token Handler.
+ */
+
+TokenHandler.prototype.handle = function(request, response) {
+  if (!(request instanceof Request)) {
+    throw new InvalidArgumentError('Invalid argument: `request` must be an instance of Request');
+  }
+
+  if (!(response instanceof Response)) {
+    throw new InvalidArgumentError('Invalid argument: `response` must be an instance of Response');
+  }
+
+  if (request.method !== 'POST') {
+    return Promise.reject(new InvalidRequestError('Invalid request: method must be POST'));
+  }
+
+  if (!request.is('application/x-www-form-urlencoded')) {
+    return Promise.reject(new InvalidRequestError('Invalid request: content must be application/x-www-form-urlencoded'));
+  }
+
+  return Promise.bind(this)
+    .then(function() {
+      return this.getClient(request, response);
+    })
+    .then(function(client) {
+      return this.handleGrantType(request, client);
+    })
+    .tap(function(data) {
+      var model = new TokenModel(data);
+      var tokenType = this.getTokenType(model);
+
+      this.updateSuccessResponse(response, tokenType);
+    }).catch(function(e) {
+      if (!(e instanceof OAuthError)) {
+        e = new ServerError(e);
+      }
+
+      this.updateErrorResponse(response, e);
+
+      throw e;
+    });
+};
+
+/**
+ * Get the client from the model.
+ */
+
+TokenHandler.prototype.getClient = function(request, response) {
+  var credentials = this.getClientCredentials(request);
+
+  if (!credentials.clientId) {
+    throw new InvalidRequestError('Missing parameter: `client_id`');
+  }
+
+  if (!credentials.clientSecret) {
+    throw new InvalidRequestError('Missing parameter: `client_secret`');
+  }
+
+  if (!is.vschar(credentials.clientId)) {
+    throw new InvalidRequestError('Invalid parameter: `client_id`');
+  }
+
+  if (!is.vschar(credentials.clientSecret)) {
+    throw new InvalidRequestError('Invalid parameter: `client_secret`');
+  }
+
+  return Promise.try(this.model.getClient, [credentials.clientId, credentials.clientSecret])
+    .then(function(client) {
+      if (!client) {
+        throw new InvalidClientError('Invalid client: client is invalid');
+      }
+
+      if (!client.grants) {
+        throw new ServerError('Server error: missing client `grants`');
+      }
+
+      if (!(client.grants instanceof Array)) {
+        throw new ServerError('Server error: `grants` must be an array');
+      }
+
```

Comment:
Already did in https://github.com/seegno-forks/node-oauth2-server/commit/995196ed108acd6804f26085240455023268ca6b.